homeassistant/components/sunweg/sensor.py

```diff
@@ -0,0 +1,173 @@
+    plant_id = config[CONF_PLANT_ID]
+
+    if plant_id == DEFAULT_PLANT_ID:
+        plant_info: list[Plant] = api.listPlants()
+        plant_id = plant_info[0].id
+
+    devices: list[Inverter] = []
+    # Get a list of devices for specified plant to add sensors for.
+    for inverter in api.plant(plant_id).inverters:
+        api.complete_inverter(inverter)
+        devices.append(inverter)
+    return [devices, plant_id]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the SunWEG sensor."""
+    name = config_entry.data[CONF_NAME]
+
+    probe: SunWEGData = hass.data[DOMAIN][config_entry.entry_id]
+
+    devices, plant_id = await hass.async_add_executor_job(
+        get_device_list, probe.api, config_entry.data
+    )
+
+    entities = [
+        SunWEGInverter(
+            probe,
+            name=f"{name} Total",
+            unique_id=f"{plant_id}-{description.key}",
+            description=description,
+            device_type="total",
+        )
+        for description in TOTAL_SENSOR_TYPES
+    ]
+
+    # Add sensors for each device in the specified plant.
+    for device in devices:
+        entities.extend(
+            [
+                SunWEGInverter(
+                    probe,
+                    name=f"{device.name}",
+                    unique_id=f"{device.sn}-{description.key}",
+                    description=description,
+                    device_type="inverter",
+                    inverter_id=device.id,
+                )
+                for description in INVERTER_SENSOR_TYPES
+            ]
+        )
+
+        for phase in device.phases:
+            entities.extend(
+                [
+                    SunWEGInverter(
+                        probe,
+                        name=f"{device.name} {phase.name}",
+                        unique_id=f"{device.sn}-{phase.name}-{description.key}",
+                        description=description,
+                        inverter_id=device.id,
+                        device_type="phase",
+                        deep_name=phase.name,
+                    )
+                    for description in PHASE_SENSOR_TYPES
+                ]
+            )
+
+        for mppt in device.mppts:
+            for string in mppt.strings:
+                entities.extend(
+                    [
+                        SunWEGInverter(
+                            probe,
+                            name=f"{device.name} {string.name}",
+                            unique_id=f"{device.sn}-{string.name}-{description.key}",
+                            description=description,
+                            inverter_id=device.id,
+                            device_type="string",
+                            deep_name=string.name,
+                        )
+                        for description in STRING_SENSOR_TYPES
+                    ]
+                )
```

Comment:
For this one you may need to reverse the order of the for loops, I am pretty sure they get evaluated in this order but if you hit an error it's because it's reversed
```suggestion
        entities.extend(
            [
                SunWEGInverter(
                    probe,
                    name=f"{device.name} {string.name}",
                    unique_id=f"{device.sn}-{string.name}-{description.key}",
                    description=description,
                    inverter_id=device.id,
                    device_type="string",
                    deep_name=string.name,
                )
                for mppt in device.mppts
                for string in mppt.strings
                for description in STRING_SENSOR_TYPES
            ]
        )
```